brew update: get the latest version of Ruby code before calling Ruby

mikemcquaid opened the pull request:
This splits `brew update` into two commands: `brew update` and `brew report`.

`brew update` handles:
- `git fetch` the latest version of the code (in parallel) and then `git merge` (or `git rebase`) the latest version and call `brew report`

`brew report` handles:
- everything involving Ruby code
- reporting the actual status of what changed between the revision changes (if any)

To accomplish this we've added a basic framework to create Homebrew commands in Bash and pushed a bit of shared logic (e.g. setting `HOMEBREW_PREFIX`) into Bash to be consistent.

I appreciate the tests aren't all there yet and the coverage isn't great; I did this to do a line-by-line port for now so people could validate the approach and then will work on getting decent test coverage.

This provides some speedup by running `git fetch` in parallel but will also pave the way for further speedup by using ETags to only do the `git fetch` when it's actually necessary to do so.

xu-cheng commented on bin/brew:
```diff
@@ -13,6 +13,8 @@ fi
 BREW_FILE_DIRECTORY="$(chdir "${0%/*}" && pwd -P)"
 HOMEBREW_BREW_FILE="$BREW_FILE_DIRECTORY/${0##*/}"
 
+export HOMEBREW_PREFIX=$(cd $(dirname $(dirname $HOMEBREW_BREW_FILE)) && pwd -P)
```
It should be `chdir` instead of `cd`

mikemcquaid replied:
:+1:

bfontaine replied:
What’s the difference?

mikemcquaid replied:
`chdir` ensures that the output is sent to `/dev/null` (in case of weird shell aliases/settings)

bfontaine replied:
Thanks. Also note that you can avoid (weird) aliases by prepending a command with `\`; e.g.:
```shellSession
$ alias cd=echo
$ cd foo
foo
$ \cd foo
bash: cd: foo: No such file or directory
```

xu-cheng commented on bin/brew:
```diff
@@ -21,7 +23,8 @@ then
 fi
 
-BREW_LIBRARY_DIRECTORY="$(chdir "$BREW_FILE_DIRECTORY"/../Library && pwd -P)"
+export HOMEBREW_REPOSITORY=$(cd $(dirname $(dirname $HOMEBREW_BREW_FILE)) && pwd -P)
```
`HOMEBREW_REPOSITORY` should based on `HOMEBREW_BREW_FILE`'s realpath. So you probably want:
```bash
HOMEBREW_REPOSITORY = "$(chdir "$BREW_FILE_DIRECTORY"/../ && pwd -P)"
```

Also, we seem to prefer to put all of `export` in the very end.

mikemcquaid replied:
:+1:

xu-cheng commented on bin/brew:
```diff
@@ -46,4 +49,16 @@ fi
-exec "$HOMEBREW_RUBY_PATH" -W0 "$BREW_LIBRARY_DIRECTORY/brew.rb" "$@"
+for i in $@
+do
+  [[ "$i" =~ "^--" ]] || continue
```
* We allow flags with one dash, e.g. `-v`. So it should be `"$i" =~ "^-"`
* `||` should be `&&`

mikemcquaid replied:
:+1:

xu-cheng commented:
Some comments:

* I would suggest to put `report.rb` outside of `cmd`. Since it shouldn't be invoked directly by users.
* I suggest to run bash script against `shellcheck`.

Some edge cases need to be addressed:

* Several checks in `brew.rb` should be implemented in bash:
   * `HOMEBREW_PREFIX` location check, [brew.rb#L51-L55](https://github.com/Homebrew/homebrew/blob/master/Library/brew.rb#L51-L55)
   * Current directory existence check, [brew.rb#L66](https://github.com/Homebrew/homebrew/blob/master/Library/brew.rb#L66)
   * Sudo check,  [brew.rb#L100-L109](https://github.com/Homebrew/homebrew/blob/master/Library/brew.rb#L100-L109)
   * Help flag check, [brew.rb#L135-L140](https://github.com/Homebrew/homebrew/blob/master/Library/brew.rb#L135-L140)
* `brew update` should check arguments and abort if formula name is passed, [update.rb#L10-L15](https://github.com/Homebrew/homebrew/blob/master/Library/Homebrew/cmd/update.rb#L10-L15)
* Current bash version of ensure git is installed will fail on non-Xcode/CLT system. Because `/usr/bin/git` stub will always exist.
* Interrupt or any failure should restore repos to original states. i.e. the original `reset_on_interrupt` should be implemented. Another problem would be handling output in this situation.

Good work though, I gotta admit I was wrong last time. :christmas_tree:

xu-cheng commented on Library/Homebrew/cmd/update.sh:
```diff
@@ -0,0 +1,204 @@
+  for i in $@
+  do
+    case "$i" in
+      --verbose) HOMEBREW_VERBOSE=1 ;;
```
We also accept `-v`. In addition, we may want to set `HOMEBREW_DEBUG`, `HOMEBREW_DEVELOPER` and `HOMEBREW_NO_COMPAT` based on flags. They will be used in `report` script.

mikemcquaid commented:
> I would suggest to put report.rb outside of cmd. Since it shouldn't be invoked directly by users.

Yeh, makes sense. I was wondering about allowing it to take e.g. SHA1/tap parameters so you can view the output for that case (which would be useful for testing).

> I suggest to run bash script against spellcheck.

You mean `shellcheck` or something here? Not quite sure I understand.

> I'm kinda curious on what will happen if two brew update instances are running simultaneously. We could potenital require a lock if it would cause any trouble.

Because we've split out `fetch` it should be fine. That said, it's maybe worth locking anyway.

> Several checks in brew.rb should be implemented in bash:

:+1: 

> Current bash version of ensure git is installed will fail on non-Xcode/CLT system. Because /usr/bin/git stub will always exist.

Yeh, I have been wondering about that case: don't we want to just encourage the users to install the CLT's `git` in that case rather than installing our own one?

> Interrupt or any failure should restore repos to original states. i.e. the original reset_on_interrupt should be implemented.

Unfortunately I don't think it's possible to ignore interrupts but we can use a `trap` here I guess to do the cleanup.

> We should avoid using brew some-command in bash script. Instead, it should be $HOMEBREW_BREW_FILE some-command. The reason is we want to avoid to pick up different brew instance on multiple installation environment.

:+1: 

> Good work though, I gotta admit I was wrong last time.

Thanks for the kind words :blush:

xu-cheng commented:
> You mean shellcheck or something here? Not quite sure I understand.

Yeah, I mean shellcheck (the original one is hit by Apple's spell correction)

> Yeh, I have been wondering about that case: don't we want to just encourage the users to install the CLT's git in that case rather than installing our own one?

I'd prefer to installing our own one. The thing is that installing CLT and accepting license requires `sudo`. And it's not friendly to automate script.

bfontaine commented on Library/Homebrew/cmd/update.sh:
```diff
@@ -0,0 +1,213 @@
+fi
+
+git_init_if_necessary() {
+  if [ -z "$(ls .git 2>/dev/null)" ]
```
You could use `[ -d .git ]` to check if `.git` exists and is a directory.

bfontaine commented on Library/Homebrew/cmd/update.sh:
```diff
@@ -0,0 +1,213 @@
+}
+
+pull() {
+  cd "$DIR" || return
```
Shouldn’t that be `chdir` given Xu’s previous comment?

mikemcquaid commented:
> Several checks in brew.rb should be implemented in bash:
> Xcode/CLT related check, otherwise git may not work. brew.rb#L25-L49
> HOMEBREW_PREFIX location check, brew.rb#L51-L55
> Current directory existence check, brew.rb#L66
> Sudo check, brew.rb#L100-L109
> Help flag check, brew.rb#L135-L140

I believe I've now addressed everything except these points. @xu-cheng let me know what you think.

xu-cheng commented on Library/Homebrew/cmd/update.sh:
```diff
@@ -0,0 +1,242 @@
+
+  if [ -z "$(which_git)" ]
+  then
+    brew install git
```
This should be `$HOMEBREW_BREW_FILE install git`

mikemcquaid replied:
Wondering if we maybe even want to declare a function or something that means that `brew install git` will use the `$HOMEBREW_BREW_FILE` as expected.

xu-cheng replied:
Good idea.

mikemcquaid replied:
Although this stayed the same it's now calling a function.

xu-cheng commented on bin/brew:
```diff
@@ -13,6 +16,8 @@ fi
 BREW_FILE_DIRECTORY="$(chdir "${0%/*}" && pwd -P)"
 HOMEBREW_BREW_FILE="$BREW_FILE_DIRECTORY/${0##*/}"
 
+HOMEBREW_PREFIX="$(chdir "$(dirname "$(dirname "$HOMEBREW_BREW_FILE")")" && pwd -P)"
```
style nit, may want to put this line in below with `HOMEBREW_REPOSITORY`, `HOMEBREW_LIBRARY`. It looks nice and more readable when putting them together,

mikemcquaid replied:
:+1:

bfontaine commented on Library/Homebrew/cmd/update.sh:
```diff
@@ -0,0 +1,242 @@
+}
+
+git_init_if_necessary() {
+  if [ -z "$(ls .git 2>/dev/null)" ]
```
You can use `[ -d .git ]` to check if `.git` is a directory.

mikemcquaid replied:
:+1:

xu-cheng commented on Library/Homebrew/cmd/update.sh:
```diff
@@ -0,0 +1,242 @@
+}
+
+update "$@"
+exec "$HOMEBREW_RUBY_PATH" -W0 "$HOMEBREW_LIBRARY/brew.rb" report "$@"
```
I suggest to put `report.rb` script outside of `cmd`. For example, you can put it in `Library/Homebrew/update_report.rb` as a standalone ruby script (similar to `build.rb`)

mikemcquaid replied:
Yeh; I'm wondering where to put it and if we want to add the support for running e.g. `brew report $SHA1 $SHA1` to be able to get the report for that diff (useful for developer testing).

xu-cheng replied:
About `brew report $SHA1 $SHA1` idea, it would be useful. But `report` may be misleading considering we also do tap and formula rename migration. It may cause some surprise.

mikemcquaid replied:
I actually think making it a standalone Ruby script may be a bad idea given all the checks in `brew.rb` that are useful. What do you think?

xu-cheng replied:
No, this is not such case. Checks in `brew.rb` should be performed before update script. So we don't need these checks in report phase.

mikemcquaid replied:
They should be but what if/when we add additional checks for things that are important and then are no longer checked before `report.rb` is run? Maybe that's not an issue, I dunno.

xu-cheng commented:
> I believe I've now addressed everything except these points

I still think these should be addressed. Especially, for sudo checker and Xcode/CLT license check.

mikemcquaid commented:
> I still think these should be addressed.

Yeh, sorry, I meant that I will be addressing them but was just running out of time so figured I'd get more review of what I'd changed.

xu-cheng commented on Library/Homebrew/cmd/update.sh:
```diff
@@ -0,0 +1,242 @@
+  do
+    case "$i" in
+      update) shift ;;
+      --verbose|-v) HOMEBREW_VERBOSE=1 ;;
```
Handle `-v` may not enough. Because we treat `-abcd` as `-a -b -c -d`. You probably want:
```bash
-*)
  for x in $(echo "$i" | grep -o .)
  do
    case "$x" in
      v) HOMEBREW_VERBOSE=1 ;;
      *);;
    esac
  done
```

mikemcquaid replied:
`--verbose|-*v*)` does what we want.

mikemcquaid replied:
(and handles spaces)

xu-cheng replied:
How about `--v` which is an incorrect flag?

xu-cheng replied:
Also `--another-flag-contains-v`?

UniqMartin replied:
To get this right, it probably needs to be something like:

```shell
case "${i}" in
  update) shift ;;
  --verbose) HOMEBREW_VERBOSE=1 ;;
  --*) : "<reject or ignore>" ;;
  -*v*) HOMEBREW_VERBOSE=1 ;;
  *) : "<fail?>" ;;
esac
```

One just has to be careful to put the cases in the right order and a bit more work has to be done if we want to check for other short flags other than `-v`.

xu-cheng commented:
Per chat, `brew update` following with immediate interrupt will cause nasty output.

chdiza commented on Library/Homebrew/cmd/update.sh:
```diff
@@ -0,0 +1,242 @@
+    then
+      which_git="$active_developer_dir/usr/bin/git"
+    fi
+    fi
```
One of these `fi` is indented too deeply.  And it looks as though the indenting of `then` is inconsistent in this function.

mikemcquaid replied:
Thanks @chdiza.

mikemcquaid commented:
I've updated this with all the comments made. I'd like to know what tests should be made here before we push this (but I'd like to try and avoid "perfect being the enemy of good" given that this will address a _lot_ of update issues for us and the earlier we ship this the easier it'll be to ship e.g. Ruby 2 support eventually).

UniqMartin commented:
I agree that we don't have to be perfect here. But we should still be confident that this hasn't any major issues (like breaking updates completely) as those are difficult to fix once this reaches a significant number of users. I'll try to test drive this and provide a timely review, but given the extent of this change and somewhat limited time next week, don't expect that to happen before the weekend …

mikemcquaid commented:
> But we should still be confident that this hasn't any major issues (like breaking updates completely) as those are difficult to fix once this reaches a significant number of users. I'll try to test drive this and provide a timely review, but given the extent of this change and somewhat limited time next week, don't expect that to happen before the weekend …

Yep, that's fine. Definitely agree with not rushing through and breaking things. That's one of the reasons why this is (more-or-less) a line-by-line port of the Ruby code into Bash. Obviously the behaviour won't be identical but the flow at least should be.

xu-cheng commented on Library/Homebrew/cmd/update.sh:
```diff
@@ -0,0 +1,259 @@
+your own risk.
+EOS
+  exit 1
+fi
```
I think we could add this check to `bin/brew` and discard the test in `brew.rb`.

mikemcquaid replied:
Yeh, I guess I didn't want to handle the "check if a file is present in this array" but I guess that's not too much pain to do.

xu-cheng commented on Library/Homebrew/cmd/update.sh:
```diff
@@ -0,0 +1,259 @@
+    fi
+  fi
+  echo $which_git
+}
```
This function seem to handle no Xcode/CLT case incorrectly. It should return nothing on such case but instead it will return `/usr/bin/git`.

xu-cheng commented on bin/brew:
```diff
@@ -45,5 +50,66 @@ fi
+then
+  echo "$HOMEBREW_PREFIX"
+  exit 0
+fi
```
We should move these code forward (i.e. before all the shell out checks)

mikemcquaid replied:
:+1:. We can also potentially have other variables handled here too.

xu-cheng replied:
If you mean `brew --prefix formula_name`, we cannot. Because it involves formula name resolution. I think `brew --prefix` may be the only command worthing the exclusive care.

mikemcquaid replied:
I was thinking perhaps `--cellar` without arguments, `--cache` without arguments and `--repository`?

mikemcquaid replied:
And maybe `--version` as I see all those commands as being things that people call from scripts that they want to have return super quickly and don't really rely on any state.

UniqMartin replied:
I believe `--cache` has the most complicated logic of these and should probably be postponed for later. For the others it make sense to me to handle them here.

mikemcquaid replied:
We also don't already have a variable for it. :+1: to the others.

xu-cheng replied:
If we want to handle `brew --repository`, we need to handle `brew --repo` alias as well.

I'm kinda :-1: for `brew --version`, whose logic is also a little complicated and may become more complicate after core formula split.

In general, I think it's wroth to write them in bash if the ruby code can be dropped. i.e. Let's avoid duplicate logic.

mikemcquaid replied:
Sounds good :+1:

xu-cheng commented on bin/brew:
```diff
@@ -45,5 +50,66 @@ fi
+for i in "$@"
+do
+  [[ "$i" =~ ^- ]] && continue
+  HOMEBREW_BASH_COMMAND="$HOMEBREW_LIBRARY/Homebrew/cmd/$i.sh"
```
I suggest to make it `cmd/$i`. So it could support any scripts in the future.

mikemcquaid replied:
I think I'd rather save that for if we want to support other scripts later and keep it more locked-down for now.

xu-cheng commented on bin/brew:
```diff
@@ -45,5 +50,66 @@ fi
+export HOMEBREW_LIBRARY
 
-exec "$HOMEBREW_RUBY_PATH" -W0 "$BREW_LIBRARY_DIRECTORY/brew.rb" "$@"
+if [ $(uname -s) = "Darwin" ]
```
There is another ` $(uname -s) = "Darwin"` before. So I think we should cache the value to avoid duplicated shell out.

UniqMartin replied:
And in any case please don't omit the quotes around `$(uname -s)` for the sake of safety and consistency.

mikemcquaid replied:
:+1:

mikemcquaid commented:
I've addressed all comments and suggestions and added tests to verify we don't accidentally have invalid Bash syntax anywhere.

This is ready for some testing and then :ship:ping. Another option for such a drastic change is we could bring back the old `update.rb` file and only run this `update.sh` if you have `HOMEBREW_DEVELOPER` set. Another another option is we push each commit here once-at-a-time and wait a bit for feedback.

Thoughts, @Homebrew/maintainers?

xu-cheng commented on Library/brew.rb:
```diff
@@ -22,32 +17,6 @@
-end
-
 case HOMEBREW_PREFIX.to_s
 when "/", "/usr"
```
I think this check should also be moved to bash script.

mikemcquaid replied:
Ok. Can you check the rest of the file to see if there's anything else? I'm just hoping to avoid too much more back-and-forth if possible because I'm not able to work on Homebrew every day at the moment.

UniqMartin replied:
I think it's sufficient if we get just the bare minimum into shell code to get `update` working. We can then make incremental changes afterwards, which will be a lot easier to review and test.

xu-cheng replied:
No problem, I'm running multiple different tests now.

xu-cheng commented on bin/brew:
```diff
@@ -21,7 +24,32 @@ then
+then
+  echo "$HOMEBREW_CELLAR"
+  exit 0
+elif [ "$*" = "--repository" ] || [ "$*" = "--repo" ]
```
You probably want:
* Remove `--repo` from `HOMEBREW_INTERNAL_COMMAND_ALIASES`.
* Make sure `brew commands` still output `--repository` and `--repo`.

bfontaine replied:
@xu-cheng How can we make `brew command --repository` working if it’s not a file anymore? What should we output?

xu-cheng replied:
I'm thinking output internal command or something. The crucial  part is to make the command return success rather than failure.

mikemcquaid replied:
I just kept `--repository.rb` as-is for now.

xu-cheng commented on bin/brew:
```diff
@@ -45,5 +78,83 @@ fi
+
+  HOMEBREW_BASH_COMMAND="$HOMEBREW_LIBRARY/Homebrew/cmd/$i.sh"
+  break
+done
```
This block can be a little confused. Would it better if we split logics among cmd argument extraction, sudo check and bash command detection?

mikemcquaid replied:
I could split them into functions, I guess? I wanted to avoid looping through the arguments multiple times but maybe it's not necessary.

xu-cheng replied:
I think the times of looping are the same.

bfontaine commented on Library/Homebrew/test/test_update_report.rb:
```diff
@@ -84,13 +78,13 @@ def test_update_homebrew_without_formulae_changes
-    assert_equal %w[xar yajl], @report.select_formula(:M)
-    assert_equal %w[antiword bash-completion ddrescue dict lua], @report.select_formula(:A)
+    # assert_equal %w[xar yajl], @report.select_formula(:M)
+    # assert_equal %w[antiword bash-completion ddrescue dict lua], @report.select_formula(:A)
```
Is there any reason you commented these out instead of removing them?

mikemcquaid replied:
Whoops, will remove them.

UniqMartin commented:
> Another option for such a drastic change is we could bring back the old `update.rb` file and only run this `update.sh` if you have `HOMEBREW_DEVELOPER` set.

I like this idea. The nature of this PR makes it a bit more difficult than usual to test while it's not yet part of the regular `master` branch. If we could force the new code on us while redirecting normal users to the old Ruby implementation, we could gather a bit more experience and if something breaks we won't have much trouble fixing our local Git repositories (unlike most of our users).

xu-cheng commented:
> The nature of this PR makes it a bit more difficult than usual to test while it's not yet part of the regular master branch.
 
@UniqMartin you can reset one of your taps back to time for test :wink:

xu-cheng commented:
Bugs found so far:

* `brew update` will overwrite custom remote to `https://github.com/Homebrew/homebrew.git`.
* If interrupt shortly after running `brew update`, it will output nasty output. We may want to set trap in prepare phase.
* `brew update` fails to report formula and tap update.

mikemcquaid commented:
> If interrupt shortly after running brew update, it will output nasty output. We may want to set trap in prepare phase.

Can you paste the output and try and see if you can propose code changes here? I've tried and failed to reproduce this locally.

> brew update fails to report formula and tap update.

Again, please can you provide e.g. the steps you ran so I can reproduce this.

xu-cheng commented on Library/Homebrew/config.rb:
```diff
@@ -31,21 +31,16 @@ def mkpath
-  HOMEBREW_REPOSITORY+"Cellar"
-end
+# Where we store built products
+HOMEBREW_PREFIX = Pathname.new(ENV["HOMEBREW_CELLAR"])
```
This should be HOMEBREW_CELLAR

xu-cheng commented:
> Can you paste the output and try and see if you can propose code changes here? I've tried and failed to reproduce this locally.

This seems to be only happening when fetching multiple taps. It will generate following output:
```
^CKilled by signal 2.
Killed by signal 2.
Killed by signal 2.
Killed by signal 2.
Killed by signal 2.
Killed by signal 2.
Killed by signal 2.
Killed by signal 2.
fatal: The remote end hung up upon initial contact
fatal: The remote end hung up upon initial contact
fatal: The remote end hung up upon initial contact
fatal: Could not read from remote repository.

Please make sure you have the correct access rights
and the repository exists.
fatal: The remote end hung up upon initial contact
```

I'm looking for a way to reproduce it more easily.

> Again, please can you provide e.g. the steps you ran so I can reproduce this.

With `homebrew/science` tapped, I set its HEAD to `80fbf89` by `git reset --hard 80fbf89`. Running `brew update`, I get `Already up-to-date`.

mikemcquaid commented:
I've addressed all comments and moved this to a separate command named `brew update-bash` that will refuse to run unless `HOMEBREW_DEVELOPER` is set. I think it would be good to focus on what's required to merge this as-is (we can iterate more on it when it's in) and if all the maintainers can use `brew update-bash` for a week or so to try and iron out edge cases.

UniqMartin commented on bin/brew:
```diff
@@ -1,7 +1,10 @@
-#!/bin/sh
+#!/bin/bash
```
Since we're now using `bash` for this file, I think it would be more idiomatic to use `[[` instead of `[` and `test` everywhere.

mikemcquaid replied:
What advantage do you get from `[[`? Doesn't feel better to me unless it's actually needed.

xu-cheng replied:
It seems that `[` will shell out.

UniqMartin replied:
All of this stuff is built-in into `bash` (try `type [`), so the performance shouldn't be an issue. As I stated, since this is now bash code, using `[[` feels more idiomatic. No other gain than that. But for me that would be enough (especially for the newly written code).

apjanke replied:
The `[[...]]` lets you use operators like `==`, `<`, and so on, which can be more readable than `-gt`, `-le`, and so on. It also gives you the `=~` pattern matching operator.

In `zsh`, `[[...]]` is syntactically safer than `[...]`, which must have the stuff inside it expanded as though it were for an external command, so it can avoid some gotchas with special characters and the like. I think the same is true for bash: in `[[...]]`, word splitting and file expansion are not done, which is probably what you want, and can save some ugly quoting and/or escaping.

UniqMartin commented on bin/brew:
```diff
@@ -1,7 +1,10 @@
 chdir() {
-  cd "$@" >/dev/null
+  cd "$@" >/dev/null || {
+    echo "Error: failed to cd to " "$@" "!" >&2
```
I think we have enough instances of “print something to `stderr` and exit with failure” to warrant a shell function, e.g.,

```shell
odie() {
  printf "%s\n" "$*" >&2
  exit 1
}
```

and then use it here like:

```shell
cd "$@" >/dev/null || odie "Error: failed to cd to $*!"
```

UniqMartin replied:
*Addendum:* Note also that `"Error: failed to cd to $*!"` feels a bit less awkward than `"Error: failed to cd to " "$@" "!"`. (And it gets rid of the superfluous white space.)

apjanke replied:
You'll probably need to escape the `!`, too. It's a special shell character for history expansion, and the expansion still happens inside double-quotes.

UniqMartin replied:
> You'll probably need to escape the `!`, too.

No, that's usually only an issue with an interactive session, but not within a script.

apjanke replied:
Ah, you're right.

UniqMartin commented on bin/brew:
```diff
@@ -21,7 +24,39 @@ then
+then
+  echo "$HOMEBREW_REPOSITORY"
+  exit 0
+fi
```
This entire block could be shortened without changing its functionality:

```shell
case "$*" in
  --prefix) echo "$HOMEBREW_PREFIX" ; exit 0 ;;
  --cellar) echo "$HOMEBREW_CELLAR" ; exit 0 ;;
  --repository|--repo) echo "$HOMEBREW_REPOSITORY" ; exit 0 ;;
esac
```

mikemcquaid replied:
:+1:

UniqMartin commented on bin/brew:
```diff
@@ -45,5 +85,77 @@ fi
+
+if [ -n "$HOMEBREW_OSX" ]
+then
+  if [ "$(xcode-\select --print-path)" = "/" ]
```
The backslash before `select` is superfluous. But we talked already about this.

UniqMartin commented on bin/brew:
```diff
@@ -45,5 +85,77 @@ fi
+then
+  if [ "$(xcode-\select --print-path)" = "/" ]
+  then
+    cat >&2 <<EOS
```
I'd suggest to replace these instances with `odie` as well, also to get rid of the `cat` call. There are two possible options. Either we keep `odie` simple as proposed further above and pass a string argument instead of a heredoc, which would look something like:

```shell
odie "\
Error text …
… across multiple lines."
```

If this looks too weird, we can keep the heredoc and extend `odie` to accept the message via arguments and `stdin` like that:

```shell
odie() {
  local message
  if [[ $# -ne 0 ]] ; then
    message="$*"$'\n'
  else
    IFS= read -rd '' message
  fi
  printf "%s" "$message" >&2
  exit 1
}

odie <<EOS
Error text …
… across multiple lines.
EOS
```

xu-cheng replied:
May I suggest to print `Error` if we implement odie. And preferably we could hightlight for `stty` to give the same output as the Ruby version.

mikemcquaid replied:
I think your `odie` there is significantly harder to follow than the heredoc. I think we could implement `odie` without the `heredoc` support and just handle it manually.

UniqMartin replied:
> I think your `odie` there is significantly harder to follow […]

Do you mean this piece of code `IFS= read -rd '' message`? Because everything else in there is trivial. I think it's not overly complicated and I think it's a good thing to use `odie` everywhere for consistency (and one won't have to remember to always add the `>&2`). As Xu says, this also makes it easier to add highlighting (which can be added later).

mikemcquaid replied:
I think it's more complex. I'm :ok: with adding `odie` for single line strings and just needing the manual `>&2` for heredocs. Again, I don't want to get too into these stylistic edge-cases in the interests of getting something merged.

UniqMartin replied:
It's slightly more complex in a single place, but avoids repetitive and error-prone code in many different places, so that's my main motivation. (Because of that, I don't think it's merely stylistic. The complex part could be explained in a comment: “read `stdin` into a variable”.

I'd be interested to know if you disagree with that in general or if you basically agree, but don't want to change it for the sake of getting something merged.

xu-cheng replied:
Fwiw, I found manually adding `>&2` can easily be forgotten. I hope we can avoid it by centralizing the implementation

xu-cheng replied:
How about we implement `odie` as such:
```bash
odie() {
    if [[ -t 1 ]]
    then
        echo -ne "\033[4;31mError\033[0m: " >&2
    else
        echo -n "Error: " >&2
    fi
    if [[ $# == 0 ]]
    then
        /bin/cat >&2
    else
        echo "$*" >&2
    fi
    exit 1
}
```

It supports all of features we talked about: highlight, heredoc, multilines and printing to stderr.

UniqMartin replied:
We need to check `stderr` with `[[ -t 2 ]]` as that's the stream we are dumping this to. Otherwise, I'd be fine with this implementation.

Rather cosmetic, but I think I would tend to replace `echo -n[e]` with `printf` (and include the colon into the highlighting) and I would use `[[ $# -eq 0 ]]` instead of `[[ $# == 0 ]]` (to get a numerical comparison; this is more about communicating intent than about the actual outcome).

mikemcquaid replied:
I'm leaving as-it is for now but will accept and review follow-up PRs after this is merged.

UniqMartin commented on bin/brew:
```diff
@@ -45,5 +85,77 @@ fi
+  XCRUN_OUTPUT="$(/usr/bin/xcrun clang 2>&1)"
+  XCRUN_STATUS="$?"
+
+  if [ "$XCRUN_STATUS" != "0" ] && [[ "$XCRUN_OUTPUT" =~ license ]]
```
Two things:

- I think it is a bit less confusing to check the exit status with a numerical comparison, i.e., `"$XCRUN_STATUS" -ne 0` instead of `"$XCRUN_STATUS" != "0"`.

- The second check doesn't really require a regular expression. Simpler and probably faster is `[[ "$XCRUN_OUTPUT" = *license* ]]`.

mikemcquaid replied:
@UniqMartin Does the globbing check not have "files in this directory named `*license*` edge cases?

UniqMartin replied:
To quote from the `bash` manual:

> Word splitting and pathname expansion are not performed on the words between the `[[` and `]]`;

`[[` isn't treated like `[` or `test`, because it's not a command, but a reserved word. (One more reason to use it.)

UniqMartin commented on bin/brew:
```diff
@@ -45,5 +85,77 @@ fi
+
+# Many Pathname operations use getwd when they shouldn't, and then throw
+# odd exceptions. Reduce our support burden by showing a user-friendly error.
+if ! [ -d "$(pwd)" ]
```
I think we should use `$PWD` instead of `$(pwd)`.

mikemcquaid replied:
`PWD` can be overriden such that it is incorrect.

UniqMartin replied:
I would hope it's not usually being overridden, but if that's a real issue, then let's keep it as-is.

UniqMartin commented on bin/brew:
```diff
@@ -45,5 +85,77 @@ fi
+# odd exceptions. Reduce our support burden by showing a user-friendly error.
+if ! [ -d "$(pwd)" ]
+then
+  echo "The current working directory doesn't exist, cannot proceed."
```
The trailing `>&2` is missing. But that's just one more instance where I'd use `odie` instead.

mikemcquaid replied:
:+1:

UniqMartin commented on bin/brew:
```diff
@@ -45,5 +85,77 @@ fi
+
+for i in "$@"
+do
+  [[ "$i" =~ ^- ]] && continue
```
We shouldn't skip arbitrary option arguments here, as this is inconsistent with the logic in `brew.rb`. For an example, compare the behavior of `brew --cache update-bash` with and without this PR. In fact, the only option that should get special treatment is `-v` and only if it's the very first argument. (There's more option-related magic with the `--help` etc. flags, but I'm not sure we should replicate the logic for that here.)

mikemcquaid replied:
I think this can be merged as-is for now as it only needs to handle `update-bash`.

UniqMartin replied:
Somewhat true. But I think it has a wider impact because it changes the rules how the command is located inside the arguments, which already happens before it's clear which command to invoke and whether that one will be implemented in shell or in Ruby.

mikemcquaid replied:
This only applies to Bash commands for now, though, so I'm going to leave it is in interests of getting this merged.

UniqMartin replied:
> This only applies to Bash commands for now, […]

Sorry to disagree again, but no. When this code is running, it's not clear yet what the command is and thus also equally unclear how it is implemented. It changes the interpretation of arguments for *everything* that is invoked via `brew`. (But I agree that most examples that show this have to be constructed and are unlikely to affect normal usage.)

mikemcquaid replied:
> (But I agree that most examples that show this have to be constructed and are unlikely to affect normal usage.)

Will leave it as-is for now.

xu-cheng replied:
Just to mention, this could cuase problem when a formula(from 3rd party or DIY) is called `update-bash`.  Or when an extrenal command starts with dash. Deep down, I would prefer it to be solved sooner if merged as it is. Leaving potential buggy code is not fun.

mikemcquaid replied:
I will ship skipping `-v` and otherwise picking the first command.

UniqMartin replied:
This is what should go somewhere before interpretation of the arguments starts:

```shell
if [[ "$1" = -v ]]
then
  shift
  set -- "$@" -v
fi
```

UniqMartin commented on bin/brew:
```diff
@@ -45,5 +85,77 @@ fi
+your own risk.
+EOS
+    exit 1
+  done
```
The entire loop can be replaced with the simpler (and probably faster):

```shell
  case "$HOMEBREW_COMMAND" in
    install|reinstall|postinstall|link|pin|unpin|update|update-bash|upgrade|create|migrate|tap|switch)
      odie <<EOS
Cowardly refusing to 'sudo brew $HOMEBREW_COMMAND'
You can use brew with sudo, but only if the brew executable is owned by root.
However, this is both not recommended and completely unsupported so do so at
your own risk.
EOS
      ;;
  esac
```

mikemcquaid replied:
:+1:

UniqMartin commented on bin/brew:
```diff
@@ -45,5 +85,77 @@ fi
+  # memory before it is run. This makes running a Bash script behave more like
+  # a Ruby script and avoids hard-to-debug issues if the Bash script is updated
+  # at the same time as being run.
+  eval "$(cat "$HOMEBREW_BASH_COMMAND")" "$@"
```
This works for `update-bash` because that one either exits via `exit` or via `exec` and never returns. To better support future shell-only commands that could simply return, I would suggest to move the remaining top-level code in the command into a function named `brew_run_command` which would look like this for `update-bash.sh`:

```shell
brew_run_command() {
  update "$@"
  exec "$HOMEBREW_BREW_FILE" update-bash-report "$@"
}
```

and then replace this line with the following two lines:

```shell
source "$HOMEBREW_BASH_COMMAND"
{ brew_run_command "$@" ; exit $? ; }
```

This will allow a command implemented in shell to return via `return <exit_code>` and also more closely matches the way internal commands are implemented in Ruby. (The grouping still makes sure that nothing else is executed, even if `bin/brew` itself is modified during an update.)

----

If you really want to stick with `eval` (I hope not), then `"$(cat "$HOMEBREW_BASH_COMMAND")"` can be simplified to `"$(<"$HOMEBREW_BASH_COMMAND")"`, i.e. no `cat` is needed.

mikemcquaid replied:
`source` could work, will try it :+1:

xu-cheng commented on Library/Homebrew/cmd/update-bash-report.rb:
```diff
@@ -0,0 +1,351 @@
+require "descriptions"
+
+module Homebrew
+  def update_bash_report
```
For the time being, let's prevent users calling these command directly. I.e. Adding homebrew developer check.

xu-cheng replied:
Also, we may want to name the command `update-report` or other name without the temporary `bash` part. Notice, we cannot change this command's name in the future.

UniqMartin replied:
That's true, we should call it `update-report` or something similar, but not use a temporary name as changing it afterwards won't be easy, if at all possible.

xu-cheng commented on Library/Homebrew/cmd/update-bash.sh:
```diff
@@ -0,0 +1,256 @@
+}
+
+git_init_if_necessary() {
+  if [ -d ".git" ]
```
This should be reversed.

UniqMartin replied:
Do we support setups where `.git` is not a directory, i.e., where the repository data is located somewhere else and `.git` is only a file that contains this directory as its contents? If so, we should only check for its existence with `-e`.

mikemcquaid replied:
We do not.

UniqMartin commented on Library/Homebrew/cmd/update-bash.sh:
```diff
@@ -0,0 +1,256 @@
+      which_git=""
+    fi
+  fi
+  echo $which_git
```
I think this needs to be quoted to avoid potential issues, e.g., with a `git` that lives in a user-provided path.

UniqMartin commented on Library/Homebrew/cmd/update-bash.sh:
```diff
@@ -0,0 +1,256 @@
+  if [ -d ".git" ]
+  then
+    git init -q
+    git config core.autocrlf false
```
Not strictly necessary, but I would prefer to be more explicit about the type of the variable, i.e., `git config --bool core.autocrlf false`.

UniqMartin commented on Library/Homebrew/cmd/update-bash.sh:
```diff
@@ -0,0 +1,256 @@
+    git init -q
+    git config core.autocrlf false
+    git config remote.origin.url https://github.com/Homebrew/homebrew.git
+    git config remote.origin.fetch +refs/heads/*:refs/remotes/origin/*
```
The refspec here needs to be quoted to avoid unintentional expansion of the `*` wildcards.

UniqMartin commented on Library/Homebrew/cmd/update-bash.sh:
```diff
@@ -0,0 +1,256 @@
+}
+
+repo_var() {
+  echo "$@" |
```
Why do we need this to expand to multiple arguments? If this function is supposed to operate on exactly one argument (I think that's the case), that should be evident from the code, i.e., `echo "$1"`.

UniqMartin commented on Library/Homebrew/cmd/update-bash.sh:
```diff
@@ -0,0 +1,256 @@
+    sed -e "s|$HOMEBREW_PREFIX||g" \
+        -e 's|Library/Taps/||g' \
+        -e 's|[^a-z0-9]|_|g' |
+    tr "[:lower:]" "[:upper:]"
```
We can always assume this to be a directory of the form `HOMEBREW_PREFIX/Library/Tap/<user>/<repo>`, right? If so, I think this is too complicated and too fragile just to extract the two last components of the path. Here's a simplified suggestion:

```shell
sed -Ee 's#^.*/([^/]+)/([^/]+)$#\1_\2#;s#[^a-z0-9]#_#g' <<< "$1" \
  | tr '[:lower:]' '[:upper:]'
```

mikemcquaid replied:
Your simplified suggestion looks significantly more complicated and less readable to me.

UniqMartin commented on Library/Homebrew/cmd/update-bash.sh:
```diff
@@ -0,0 +1,256 @@
+upstream_branch() {
+  local upstream_branch
+  upstream_branch="$(git symbolic-ref refs/remotes/origin/HEAD 2>/dev/null |
+                     sed -e 's|refs/remotes/origin/||' )"
```
I suppose the pattern should be anchored at the beginning. Better yet, let's do more stuff directly in `bash` instead of spawning external commands all the time, especially for simple tasks like trimming a prefix from a string, which could be done like this:

```shell
upstream_branch="$(git symbolic-ref refs/remotes/origin/HEAD 2>/dev/null)"
upstream_branch="${upstream_branch#refs/remotes/origin/}"
```

UniqMartin commented on Library/Homebrew/cmd/update-bash.sh:
```diff
@@ -0,0 +1,256 @@
+  # ensure GIT_CONFIG is unset as we need to operate on .git/config
+  unset GIT_CONFIG
+
+  cd "$HOMEBREW_REPOSITORY" || {
```
I think we can use `chdir` from `bin/brew` here instead of replicating the logic.

UniqMartin commented on Library/Homebrew/cmd/update-bash.sh:
```diff
@@ -0,0 +1,256 @@
+}
+
+update "$@"
+exec brew update-bash-report "$@"
```
This needs to be `exec "$HOMEBREW_BREW_FILE" update-bash-report "$@"` because `exec` cannot execute a shell function (and thus doesn't check if one exists).

UniqMartin commented on Library/Homebrew/cmd/update-bash.sh:
```diff
@@ -0,0 +1,256 @@
+    pull "$DIR"
+  done
+
+  cd "$HOMEBREW_REPOSITORY" || {
```
I think we can use `chdir` from `bin/brew` here instead of replicating the logic.

UniqMartin commented on Library/Homebrew/cmd/update-bash.sh:
```diff
@@ -0,0 +1,256 @@
+  for DIR in $HOMEBREW_REPOSITORY $HOMEBREW_LIBRARY/Taps/*/*
+  do
+    [ -d "$DIR/.git" ] || continue
+    pull "$DIR"
```
This argument is never used in `pull`, but it instead uses the global variable `DIR`. At minimum, this arguments needs to be removed, to avoid confusion.

My preferred solution of course would be, that `DIR` becomes a local variable in `update` and `pull` and that the latter extracts it from its arguments, i.e., starts with `local DIR="$1"`.

UniqMartin commented on Library/Homebrew/cmd/update-bash.sh:
```diff
@@ -0,0 +1,256 @@
+  }
+  git_init_if_necessary
+
+  for DIR in "$HOMEBREW_REPOSITORY" $HOMEBREW_LIBRARY/Taps/*/*
```
`$HOMEBREW_LIBRARY` should be quoted, like any other variable, except in the extremely rare cases where we want to have expansion. This line should preferably look like this:

```shell
for DIR in "$HOMEBREW_REPOSITORY" "$HOMEBREW_LIBRARY"/Taps/*/*
```

UniqMartin commented on Library/Homebrew/cmd/update-bash.sh:
```diff
@@ -0,0 +1,256 @@
+
+  wait
+
+  for DIR in $HOMEBREW_REPOSITORY $HOMEBREW_LIBRARY/Taps/*/*
```
Both variables need to be quoted:

```shell
for DIR in "$HOMEBREW_REPOSITORY" "$HOMEBREW_LIBRARY"/Taps/*/*
```

UniqMartin commented on Library/Homebrew/cmd/update-bash.sh:
```diff
@@ -0,0 +1,256 @@
+
+  for DIR in "$HOMEBREW_REPOSITORY" $HOMEBREW_LIBRARY/Taps/*/*
+  do
+    [ -d "$DIR/.git" ] || continue
```
Because `.git` can be a file (like stated in another comment), maybe prefer `-e`?

UniqMartin commented on Library/Homebrew/cmd/update-bash.sh:
```diff
@@ -0,0 +1,256 @@
+
+  for DIR in $HOMEBREW_REPOSITORY $HOMEBREW_LIBRARY/Taps/*/*
+  do
+    [ -d "$DIR/.git" ] || continue
```
Because `.git` can be a file (like stated in another comment), maybe prefer `-e`?

UniqMartin commented on Library/Homebrew/cmd/update-bash.sh:
```diff
@@ -0,0 +1,256 @@
+    [ -d "$DIR/.git" ] || continue
+    cd "$DIR" || continue
+    TAP_VAR=$(repo_var "$DIR")
+    export HOMEBREW_UPDATE_BEFORE"$TAP_VAR"="$(git rev-parse "$(upstream_branch)")"
```
Thinking a bit more about this, maybe we could avoid the complications of `repo_var` completely and avoid having to set a bunch of different environment variables. The idea I have in mind:

```shell
HOMEBREW_UPDATE_BEFORE+=":$DIR:$(git rev-parse "$(upstream_branch)")"
```

and once we iterated over all of them, export it. The Ruby part won't have much difficulty extracting the colon-separated directory-revision pairs from this string.

(We would do a similar thing for `HOMEBREW_UPDATE_AFTER` in `pull`.)

mikemcquaid replied:
I prefer the current implementation.

UniqMartin replied:
Why? IMHO, it is substantially more complicated (it needs to generate a valid and unique environment variable from a directory name) while the proposed change makes life a lot simpler (just appending strings).

mikemcquaid replied:
I prefer having separate environment variables; it makes it easier to inspect and debug. I don't trust having a string where you're reading it off a pair at a time and every second value has a different meaning.

mikemcquaid replied:
But let's not get into this now; as mentioned: this is getting merged soon so let's save discussion for actual bugs/problems rather than stylistic issues.

UniqMartin commented on Library/Homebrew/cmd/update-bash.sh:
```diff
@@ -0,0 +1,256 @@
+      exit 1
+    fi
+    return
+  fi
```
The `--simulate-from-current-branch` feature is completely broken because `$CURRENT_REVISION` is never passed on to the command responsible for the reporting. (Also, `git fetch` will still run in this mode, though it doesn't need to – and probably shouldn't for efficiency reasons.)

mikemcquaid replied:
Now that this is a separate command I'm OK with making this something that can be addressed in a future PR.

UniqMartin commented on Library/Homebrew/cmd/update-bash.sh:
```diff
@@ -0,0 +1,256 @@
+  # origin branch name is, and use that. If not set, fall back to "master".
+  INITIAL_BRANCH="$(git symbolic-ref --short HEAD 2>/dev/null)"
+  UPSTREAM_BRANCH="$(upstream_branch)"
+  export HOMEBREW_UPDATE_AFTER"$TAP_VAR"="$(git rev-parse "$UPSTREAM_BRANCH")"
```
Why is this set so early in `pull`? Shouldn't this be postponed to the point where `$CURRENT_REVISION` is set for the last time? (Haven't tested this yet, but it seems it will contain the revision from before the merge, so the report later on will be always empty?!)

UniqMartin commented on Library/Homebrew/cmd/update-bash.sh:
```diff
@@ -0,0 +1,256 @@
+  if [ "$INITIAL_BRANCH" != "$UPSTREAM_BRANCH" ] && [ -n "$INITIAL_BRANCH" ]
+  then
+    git checkout --force -B "$UPSTREAM_BRANCH" "origin/$UPSTREAM_BRANCH" $QUIET_ARGS
+  fi
```
This piece doesn't take into account the changes made in #46951 (probably my fault, as I merged it after this PR was created).

mikemcquaid replied:
:+1:

UniqMartin commented on Library/Homebrew/cmd/update-bash.sh:
```diff
@@ -0,0 +1,256 @@
+    brew install git
+    if [ -z "$(which_git)" ]
+    then
+      echo "Error: Git must be installed and in your PATH!" >&2
```
There's an `exit 1` missing after this line. Failing to find/install `git` won't stop the command.

UniqMartin commented on Library/Homebrew/cmd/update-bash.sh:
```diff
@@ -0,0 +1,256 @@
+
+  if ! test -w "$HOMEBREW_REPOSITORY"
+  then
+    echo "Error: $HOMEBREW_REPOSITORY must writable!" >&2
```
The verb `be` is missing in this message.

UniqMartin commented on Library/Homebrew/cmd/update-bash.sh:
```diff
@@ -0,0 +1,256 @@
+  # check permissions
+  if [ "$HOMEBREW_PREFIX" = "/usr/local" ] && ! test -w /usr/local
+  then
+    echo "Error: /usr/local must writable!" >&2
```
The verb `be` is missing in this message.

UniqMartin commented on Library/Homebrew/cmd/update-bash.sh:
```diff
@@ -0,0 +1,256 @@
+      --simulate-from-current-branch) HOMEBREW_SIMULATE_FROM_CURRENT_BRANCH=1 ;;
+      --*) ;;
+      -*v*) HOMEBREW_VERBOSE=1 ;;
+      *)
```
There's a case missing directly before this one, namely to handle other single-letter options: `-*) ;;`. Currently, `brew update-bash -d` results in:

```
This command updates brew itself, and does not take formula names.
Use 'brew upgrade <formula>'.
```

UniqMartin commented on Library/Homebrew/cmd/update-bash.sh:
```diff
@@ -0,0 +1,256 @@
+  do
+    case "$i" in
+      update|update-bash) shift ;;
+      --help) exec brew update --help ;;
```
This is also not going to call the `brew` function defined in this file because of the `exec`.

UniqMartin commented on Library/Homebrew/cmd/update-bash.sh:
```diff
@@ -0,0 +1,256 @@
+    exit 1
+  fi
+
+  for i in "$@"
```
Purely stylistic, but `in $@` can be omitted when iterating over the elements of `$@` (this is implicitly assumed).

bfontaine replied:
@UniqMartin Do you mean we can use `for i` only? I think explicit is better than implicit here.

UniqMartin replied:
Yes, exactly. But I agree that being explicit is less cryptic in this case.

mikemcquaid commented:
(Most) comments addressed, tested and pushed. Thanks for all the review, folks.
